File: backend/app/routers/lyrics.py
import httpx
from fastapi import APIRouter, Query, HTTPException
from typing import Dict, Any, Optional
from app.services.cache_service import search_cache

router = APIRouter(prefix="/api/lyrics", tags=["Lyrics"])

@router.get("", response_model=Dict[str, Any])
async def get_lyrics(
    track: str = Query(..., min_length=1),
    artist: str = Query("", description="Artist name")
):
    clean_track = track.strip()
    clean_artist = artist.strip()
    cache_key = f"lyrics:{clean_track.lower()}:{clean_artist.lower()}"
    cached = search_cache.get(cache_key)
    if cached:
        return cached

    try:
        async with httpx.AsyncClient(timeout=4.0) as client:
            params = {"track_name": clean_track}
            if clean_artist:
                params["artist_name"] = clean_artist

            res = await client.get("https://lrclib.net/api/get", params=params)
            if res.status_code == 200:
                data = res.json()
                result = {
                    "plain_lyrics": data.get("plainLyrics", ""),
                    "synced_lyrics": data.get("syncedLyrics", ""),
                    "found": True
                }
                search_cache.set(cache_key, result, ttl=86400)
                return result

            # Search fallback
            search_res = await client.get("https://lrclib.net/api/search", params={"q": f"{clean_artist} {clean_track}".strip()})
            if search_res.status_code == 200:
                items = search_res.json()
                if isinstance(items, list) and items:
                    item = items[0]
                    result = {
                        "plain_lyrics": item.get("plainLyrics", ""),
                        "synced_lyrics": item.get("syncedLyrics", ""),
                        "found": True
                    }
                    search_cache.set(cache_key, result, ttl=86400)
                    return result
    except Exception as err:
        print(f"Lyrics lookup error: {err}")

    result = {
        "plain_lyrics": f"♪ {clean_track} ♪\nby {clean_artist or 'Unknown Artist'}\n\n(Instrumental / Lyrics unavailable)",
        "synced_lyrics": "",
        "found": False
    }
    return result

File: frontend/src/services/lyrics.js
import api from './api';

export const getLyrics = async (track, artist = '') => {
  try {
    const response = await api.get(`/lyrics?track=${encodeURIComponent(track)}&artist=${encodeURIComponent(artist)}`);
    if (response.data && (response.data.plain_lyrics || response.data.synced_lyrics)) {
      return response.data;
    }
  } catch (err) {
    console.warn('Backend lyrics fetch failed, trying public LRCLIB API...', err);
  }

  // Public fallback via LRCLIB (free open lyrics database)
  try {
    const searchUrl = `https://lrclib.net/api/search?q=${encodeURIComponent(track + ' ' + artist)}`;
    const res = await fetch(searchUrl);
    if (res.ok) {
      const list = await res.json();
      if (Array.isArray(list) && list.length > 0) {
        const first = list[0];
        return {
          plain_lyrics: first.plainLyrics || `♪ ${track} ♪\nby ${artist || 'Unknown Artist'}\n\n(No text lyrics found)`,
          synced_lyrics: first.syncedLyrics || '',
          found: true,
        };
      }
    }
  } catch (lrcErr) {
    console.warn('LRCLIB fallback failed:', lrcErr);
  }

  return {
    plain_lyrics: `♪ ${track} ♪\nby ${artist || 'Unknown Artist'}\n\n(Lyrics unavailable)`,
    synced_lyrics: '',
    found: false,
  };
};
